### PR TITLE
[whowatch] add support for whowatch.tv (closes #223)

### LINKED ISSUE #223: [Site Request] https://whowatch.tv/
## Checklist

- [x] I'm reporting a new site support request
- [x] I've verified that I'm running yt-dlp version **2021.04.03**
- [x] I've checked that all provided URLs are alive and playable in a browser
- [x] I've checked that none of provided URLs violate any copyrights
- [x] I've searched the bugtracker for similar site support requests including closed ones

## Example URLs

- Single live: https://whowatch.tv/viewer/20955774 (might be offline by the time checked)
- https://whowatch.tv/ <- main page able to grab anyone online at the time to check

## Description

Japanese livestream website.

### PR DESCRIPTION
**Pull request by nao20010128nao**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

For note, `whowatch.py` has been taken from https://github.com/nao20010128nao/ytdl-patched/blob/9e4a0e061a558cdb05a618e27f47ca0ac56ece94/youtube_dl/extractor/whowatch.py , which is my own work.

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information
This fixes #223. 

Btw I have a question regarding this PR, please check Discord.

### CODE REVIEW

**Review by pukkandan — COMMENTED** [review]
1. I see fields like `started_at`, `total_view_count`, `comment_count` and `comments` in the API response. Since we can only download the video while it is streaming, I am not sure how usefull most of these fields will be, but `started_at` atleast is usefull as `timestamp`
1. The formats have a `quality` which should be extracted so that even if we fail to get resolution, sorting can still work
1. `uploader_id_internal` is not a valid field. `id` should be converted to string and used as `uploader_id` while `user_path` should be converted to `uploader_url` by prefixing `https://whowatch.tv/profile/`
1. The new protocol must be documented, https://github.com/yt-dlp/yt-dlp/blob/eb55bad5a0c1af9388301ffbf17845ee53a41635/yt_dlp/extractor/common.py#L158-L161 added to `SUPPORTED_PROTOCOLS` for `ffmpeg`, https://github.com/yt-dlp/yt-dlp/blob/eb55bad5a0c1af9388301ffbf17845ee53a41635/yt_dlp/downloader/external.py#L341 and shortened names should be defined. https://github.com/yt-dlp/yt-dlp/blob/eb55bad5a0c1af9388301ffbf17845ee53a41635/yt_dlp/downloader/__init__.py#L46-L60 I can do these if you want though

**Comment by nao20010128nao** [comment]
Done.

File: yt_dlp/extractor/whowatch.py [hunk 1]
```diff
@@ -67,18 +73,28 @@ def _real_extract(self, url):
...
-        uploader_id_internal = try_get(metadata, lambda x: x['live']['user']['id'], int)
+        uploader_url = try_get(metadata, lambda x: x['live']['user']['user_path'], compat_str)
+        if uploader_url:
+            uploader_url = 'https://whowatch.tv/profile/%s' % uploader_url
+        uploader_id = compat_str(try_get(metadata, lambda x: x['live']['user']['id'], int))
         uploader = try_get(metadata, lambda x: x['live']['user']['name'], compat_str)
         thumbnail = try_get(metadata, lambda x: x['live']['latest_thumbnail_url'], compat_str)
+        upload_date = try_get(metadata, lambda x: x['live']['started_at'], int)
+        if upload_date is not None:
+            upload_date = datetime.datetime.fromtimestamp(upload_date / 1000).strftime('%Y%m%d')
```
Comment (pukkandan) [inline yt_dlp/extractor/whowatch.py]:
```suggestion
        timestamp = int_or_none(try_get(metadata, lambda x: x['live']['started_at']), scale=1000)
```

see below

File: yt_dlp/extractor/whowatch.py [hunk 2]
```diff
@@ -67,18 +73,28 @@ def _real_extract(self, url):
...
         return {
             'id': video_id,
             'title': title,
             'uploader_id': uploader_id,
-            'uploader_id_internal': uploader_id_internal,
+            'uploader_url': uploader_url,
             'uploader': uploader,
             'formats': formats,
             'thumbnail': thumbnail,
+            'upload_date': upload_date,
```
Comment (pukkandan) [inline yt_dlp/extractor/whowatch.py]: `upload_date` will be automatically calculated from `timestamp`

```suggestion
            'timestamp': timestamp,
```

File: yt_dlp/extractor/whowatch.py (hunk 1 above)
Comment (nao20010128nao) [reply yt_dlp/extractor/whowatch.py]: done

**Comment by nao20010128nao** [comment]
For defining shortened names, should I add `'rtmp_ffmpeg': 'rtmp',` for that?

File: yt_dlp/extractor/whowatch.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+                    'vcodec': 'h264',
+                    'acodec': 'aac',
+                    'quality': quality,
+                    'format_note': fmt.get('label'),
+                    # note: HLS and RTMP have same resolution for now, so it's acceptable
+                    'width': try_get(hls_fmts, lambda x: x[0]['width'], int),
+                    'height': try_get(hls_fmts, lambda x: x[0]['height'], int),
+                })
+
+        formats.extend(self._extract_m3u8_formats(
```
Comment (pukkandan) [inline yt_dlp/extractor/whowatch.py:71]:
```suggestion
        # This contains the same formats as the above manifests and is used only as a fallback
        formats.extend(self._extract_m3u8_formats(
```

**Pull request merged**